### PR TITLE
introduce interview process for operations background team members

### PR DESCRIPTION
**Pull request by knisbet**
- Proposes new interview challenge for candidates with backgrounds in various operations disciplines (DevOps, System Administrators, etc)
- Create leveling guide for operations disciplines
- Create lightweight automation challenge and port existing production troubleshooting interview for team members

TODO:
- I'll port to PDF versions prior to merge
-

### CODE REVIEW

File: challenges/platform/README.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
+# Summary
+
+Teleport Cloud provides secure access to servers, databases, and applications of our customers. The platform team is responsible for automating, scaling, monitoring, and securing the infrastructure that underpins Teleport Cloud. This team draws on a wide variety of skills and experience to build reliable infra, and as such we use different interviews based on the candidates background.
```
Comment (ulysseskan) [inline challenges/platform/README.md:3]: "candidate's background"

**Review by ulysseskan — COMMENTED** [review]
added one tiny grammar suggestion

**Comment by NajiObeid** [comment]
Looking that the repo doesn't really have guidelines for format (tex vs md vs something else) or structure, I wouldn't want to comment on that. I guess at some point someone will have to take care of all that as the company grows and more roles are introduced.

My main concern is the balance between interviews for systems and operations engineers. I am sure it's been mentioned before during team meetings but seeing it in writing worries me a bit. What I mean is that this would be seen as the easier path and more people would be going for it, especially for the lower level people that are still unsure about their career paths or don't mind changing.. That being said I have no idea what problems will be proposed in the troubleshooting session and it might balance out at the end.

Another concern would be the time investment. Just comparing level 1 of both systems and operations interviews show that there is a big imbalance. I am not sure how or if we can avoid that.

**Comment by knisbet** [comment]
> Looking that the repo doesn't really have guidelines for format (tex vs md vs something else) or structure, I wouldn't want to comment on that. I guess at some point someone will have to take care of all that as the company grows and more roles are introduced.

Why does that matter? I'm more familiar with markdown, so for me it's a faster and easier way to publish an interview guide. For others they may prefer working in tex, they can use tex. 

> My main concern is the balance between interviews for systems and operations engineers. I am sure it's been mentioned before during team meetings but seeing it in writing worries me a bit. What I mean is that this would be seen as the easier path and more people would be going for it, especially for the lower level people that are still unsure about their career paths or don't mind changing.

I suspect the part I may have failed to convey, is while it's hiring for the same team, I'm treating this as a separate role, which is why it has it's own leveling guide, even though there is probably alot of cross over. So someone going through this interview process would be applying for a different posting, with a different set of skill requirements, and a different set of expected contributions to the team. This is why I wrote a separate leveling guide, which may have different comp tied to it based on market rates.

What's important, is that the interview process produces signal that we can use to understand who is a strong candidate for the team and contribute to the more operationally focused parts of the team we're currently lacking on, not how much time it takes to complete the interview.

**Comment by NajiObeid** [comment]
I agree that what matters is hiring the right people.

What I am pointing out is if we can get a good signal in this style of interviews that may mean that the other process is unnecessarily demanding. We should try to eliminate any bias applicants might have when faced with two options, engineering or ops. I can guarantee that some people will go for the ops role only because the interview process "seems" easier and less demanding on paper. Having all this info available publicly does not help. I am not saying we should change anything about any of these processes now, it's just something I noticed and wanted to mention.

File: challenges/platform/devops.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+Implement all of the level 1 requirements plus:
+1. Add a prometheus endpoint to report metrics on the following:
+    1. Counter - number of new connections
+
+    Tip: There are client libraries available in many languages for providing prometheus metrics, see https://prometheus.io/docs/instrumenting/clientlibs/
+2. Build the project into a docker container, provide instructions on how to execute as a container but reporting connections on the host.
+
+### Questions
+1. If you had to deploy this program to hundreds of servers, what would be your preferred method? Why?
+2. Describe a problem and how you solved that problem that you are particularly proud of?
```
Comment (dboslee) [inline challenges/platform/devops.md:87]: I would consider moving this to Level 2 and moving up the following to Level 3
```4. What is the hardest technical problem or outage you've had to solve in your career? Explain what made it so difficult?```

File: challenges/platform/README.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+1. [Devops development challenge](devops.md)
+2. [Troubleshooting challenge](troubleshooting.md)
+
+### Levels
+
+
+| Level  | Oncall  | Operations  | Security  | Communications / Writing  | Networking  | Systems Engineering  | OS Understanding  | Tooling |
+|---|---|---|---|---|---|---|---|---|
+| 1 | Participates in on-call and able to execute run books | Documents and Executes maintenance activities in production | Able to follow security guidelines and best practices in a production environment  | Contributor to internal documentation such as updating run books | Able to troubleshoot introductory networking issues, network policies, routing, etc. | Focus on and contribute to specific area of production system (such as the deployment of our monitoring stack)  | Working understanding of core OS concepts, filesystem, networking, cgroups, namespaces, security mechanisms, etc.  | Work with team members on deployment and maintenance of internal tooling |
```
Comment (wadells) [inline challenges/platform/README.md:44]: Do we want to have our entry level participate in on-call?  I'd argue we should have on-call start at l2 -- with the idea that entry level may have 0 industry experience, and thus probably need at least 6m-1y of experience and shadowing before they'd be ready to participate as a first class member.

I say this because I don't want our entry level to actually be "1-2 years of experience"  -- it stops being entry level -- and l2 on-call also gives us a carrot to entice entry level towards on-call.

Under the theory of "you need to perform at the level before you're promoted to it", the upper end of l1 can start participating in the rotation in preparation for promotion to l2.

Language suggestion:

l1 Shadows on-call and executes run books
l2 Serves as primary on call, able to investigate and solve many production incidents with peer review

File: challenges/platform/README.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+| 3 | Provide leadership, guidance, and act as escalation point for junior team members |   | Take charge of security incident coordination and create recommendations based on new findings | Contribute to engineering blog, meetups, booth duty, etc to advertise product and engineering culture at Teleport  | Solid understanding of advanced networking, routing, load balancing concepts and scaling production systems  | Design and implement significant automation for challenging problem spaces, such as implementing horizontal/vertical autoscaling and resource allocation | Solid understanding of workload isolation, syscall filtering, linux internals, for high security and production workloads  | Take design lead on key operational areas, such as introducing new tooling, new monitoring, or complex automation |
```
Comment (wadells) [inline challenges/platform/README.md:46]: Security:
> Take charge of security incident coordination and create recommendations based on new findings

I want "create recommendations" to be a bit more active ownership at l3.  No "I recommend others do the work" -- we want folks to drive what they believe in to done/good enough.  Perhaps:

> Take charge of security incident coordination, create and execute improvements based on new findings

File: challenges/platform/README.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+TBD - Use Devops challenge for now
+
+## Cloud Engineering
+New Challenge TBD
+
+## DevSecOps
+TBD - Use Devops challenge for now, security challenge TBD
+
+## Security Engineering
+TBD
```
Comment (wadells) [inline challenges/platform/README.md:59]: YAGNI -- lets not hit at or promise things before they're ready -- it presents a less professional front.

File: challenges/platform/README.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+Systems developers on the platform team leverage a strong background in software development to build, maintain, deploy core production services.
+
+### Challenge
+1. [Systems developer challenge](../systems/worker.pdf)
+
+### Levels
+See [levels.pdf](../../levels.pdf)
+
+
+## Devops / DevSecOps / System Administrators / Production engineers / etc
```
Comment (wadells) [inline challenges/platform/README.md:30]: s/System Administrators/SREs/ ?

My understanding is we're not presently looking to hire anyone _seeking_ a System Administrator role, but it is absolutely great to have folks with sysadmin in their background if they're seeking for a devops/infrastructure as code role.

**Review by wadells — COMMENTED** [review]
> What I am pointing out is if we can get a good signal in this style of interviews that may mean that the other process is unnecessarily demanding.

I wouldn't say unnecessarily demanding overall, so much as focused on the wrong area.  Folks who can complete the linux job worker challenge have great dev skills for our team, but may have their expectations about the nature of our work misset, and that challenge does little to show off their e.g networking, or troubleshooting skills.  I don't see that as less demanding so much as a different focus.

File: challenges/platform/devops.md (hunk 2 above)
Comment (wadells) [reply challenges/platform/devops.md:87]: Or cut one of these two.  I'm not sure we get that much extra signal by having both?

File: challenges/platform/devops.md [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+## Level 2:
+Implement all of the level 1 requirements plus:
+1. Add a Makefile or your preferred build scripting to build and test the solution.
+2. Add some tests, test for corner cases or unexpected behaviour.
+
+### Questions
+1. Why did you choose `x` to write the build automation?
+2. Is there anything else you would test if you had more time?
+3. What is the most important tool, script, or technique you have for solving problems in production? Explain why this tool/script/technique is the most important.
+4. What is the hardest technical problem or outage you've had to solve in your career? Explain what made it so difficult?
```
Comment (wadells) [inline challenges/platform/devops.md:75]: I'm concerned that we may see more subjectivity and bias in our interview process with questions like the following:

> 3. What is the most important tool, script, or technique you have for solving problems in production? Explain why this tool/script/technique is the most important.
> 4. What is the hardest technical problem or outage you've had to solve in your career? Explain what made it so difficult?

The risk here being we favor a monoculture of ideas and experiences E.g. Are we going to favor candidates that like tools we already like (as opposed to interesting new things we may have never used).  Could toughest challenge become pattern matching for "people who solved the same challenges as me"? However I can also see why they're useful (coming from a long background of behavioral interviewing).

What if we added a candidate guideline:

> Please answer the following questions in writing with a few bullet points up to a few sentences (<100 words).

I hope this will:
 - keep the amount of work reasonable for candidates
 - cut out potential sources of bias (long answers vs short)

Maybe we have our internal scoring rubrick have some language about [exploration vs exploitation](https://en.wikipedia.org/wiki/Multi-armed_bandit#Empirical_motivation).

File: challenges/platform/README.md (hunk 3 above)
Comment (knisbet) [reply challenges/platform/README.md:44]: So my line of thinking when writing this, is that a new grad / entry level would be expected to be oncall relatively quickly, with backup / escalation to a senior engineer. So if there is a prescribed runbook, they can just execute it and get experience. For current state we might have too many risks in our architecture to really throw a new grad the keys to prod, but eventually we'll get to the state... unless we get to the point where we need to double up to get admin on prod.

This is how I got started, I was oncall for a key piece of network infra within a couple months of graduating.

File: challenges/platform/README.md (hunk 6 above)
Comment (knisbet) [reply challenges/platform/README.md:30]: Yea, it's really tricky, oldschool sysadmins probably did all of these requirements and more, but got a bad reputation, so we starting calling them devops, which now has just as much of a bad reputation as companies started calling their admin teams DevOps, so now they're called SREs, and now SREs have a bad rep in some circles..

File: challenges/platform/devops.md (hunk 7 above)
Comment (knisbet) [reply challenges/platform/devops.md:75]: > The risk here being we favor a monoculture of ideas and experiences. E.g. Are we going to favor candidates that like tools we already like (as opposed to interesting new things we may have never used). Could toughest challenge become pattern matching for "people who solved the same challenges as me"? However I can also see why they're useful (coming from a long background of behavioral interviewing).

This is very true, but I don't know is inherent to the question or the person reviewing the answer. Basically I'm trying to introduce some questions that demonstrate intellectual curiosity and solid reasoning. Any solution is appropriate if it can be explained, which is why I'm trying to keep the question open and vague. And the candidate is able to self select an area of their own experience, not directed to something we're good at (like how we used to ask systems engineers to design authentication/authorization, and the only right answer was mTLS).

File: challenges/platform/README.md (hunk 6 above)
Comment (knisbet) [reply challenges/platform/README.md]: I tried to target these more to focus areas, and then include the example list. LMK if that comes across a bit better.

**Comment by knisbet** [comment]
> I agree that what matters is hiring the right people.
> 
> What I am pointing out is if we can get a good signal in this style of interviews that may mean that the other process is unnecessarily demanding. We should try to eliminate any bias applicants might have when faced with two options, engineering or ops. I can guarantee that some people will go for the ops role only because the interview process "seems" easier and less demanding on paper. Having all this info publicly available does not help. I am not saying we should change anything about any of these processes now, it's just something I noticed and wanted to mention.

@NajiObeid Thanks for noticing, I'll keep it in mind, and try and address with Tommy that resumes with mismatched backgrounds get flagged. Otherwise, I'm open to suggestions.

**Comment by knisbet** [comment]
@wadells / @dboslee / @NajiObeid / @ulysseskan / @bernardjkim 

I took another pass and adjusted a few things:
1. Added some scope for L2/L3 based on my discussion with Sasha, to also detect port scans. I'm a bit worried that this is just scope creeping, so LMK what you think.
2. Added a note on the tips section about answer length
3. Removed the duplicate question
4. Fixed typos and grammar
5. Adopted walts suggestions in the leveling guide (except for oncall for L1, feel free to jump on the discussion)

File: challenges/platform/README.md (hunk 3 above)
Comment (wadells) [reply challenges/platform/README.md:44]: I think we're saying the same thing with different language.  My "shadow" is your "oncall with backup".  The main distinction for me is: can I take a hike without my phone during the shift?  If the answer is "no" then the mentee is still in some phase of learning, and not full on-call.

Comment (wadells) [reply challenges/platform/README.md:44]: Perhaps:

l1:
> Participates in on-call with backup...

File: challenges/platform/devops.md [hunk 8]
```diff
@@ -0,0 +1,122 @@
+# Summary
```
Comment (klizhentas) [inline challenges/platform/devops.md:1]: The challenge with levels looks good to me - I like the combination of troubleshooting + scripting, a good mix.

File: challenges/platform/README.md [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+1. [Devops development challenge](devops.md)
+2. [Troubleshooting challenge](troubleshooting.md)
+
+### Levels
+
+
+| Level  | Oncall  | Operations  | Security  | Communications / Writing  | Networking  | Systems Engineering  | OS Understanding  | Tooling |
+|---|---|---|---|---|---|---|---|---|
+| 1 | Participates in on-call and able to execute run books | Documents and Executes maintenance activities in production | Able to follow security guidelines and best practices in a production environment  | Contributor to internal documentation such as updating run books | Able to troubleshoot introductory networking issues, network policies, routing, etc. | Focus on and contribute to specific area of production system (such as the deployment of our monitoring stack)  | Working understanding of core OS concepts, filesystem, networking, cgroups, namespaces, security mechanisms, etc.  | Work with team members on deployment and maintenance of internal tooling |
```
Comment (klizhentas) [inline challenges/platform/README.md:44]: I would zoom in here on skills and technologies that are relevant to the team. This leveling does not have to be universally valid, just valid for the cloud team. For example, as a L1 engineer, reading this level, I should ideally get a good answer, what technology and skill should I understand and practice to get to L2 and how do I demonstrate that? 

For example, I would elaborate on:

Solid understanding of advanced networking, routing, load balancing concepts and scaling production systems

We probably mean something quite specific here, like container networking, ability to set-up VPNs and tunnels, traffic routing, etc. Someone reading this would then understand what's required and how to get there.

**Review by klizhentas — COMMENTED** [review]
Overall looks good to me, just have a couple of tweaks to levels - the more specific we can get, the better. I think as you get more experience with candidate's submissions, you folks can tweak the challenges and levels as well.

**Comment by klizhentas** [comment]
@NajiObeid - re concerns about showing this information and someone gaming the system. In our interview guides we mention that our interview process is non-adversary by design - we try to set up the candidate for success and give them all the tools to demonstrate their best abilities. In my experience I haven't encountered anyone yet who used our transparency to game the system. Theoretically there could be a case of someone tricking us, but I'd take this risk to keep everyone's interview experience as open as possible.

File: challenges/platform/README.md (hunk 9 above)
Comment (knisbet) [reply challenges/platform/README.md]: I rotated the table so hopefully it formats better and did a bit of tuning to the language to include more examples, but probably requires more work still.

File: challenges/platform/devops.md [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+
+    Tip: There are client libraries available in many languages for providing prometheus metrics, see https://prometheus.io/docs/instrumenting/clientlibs/
+2. Build the project into a docker container, provide instructions on how to execute as a container while reporting connections on the host.
+3. When a port scan is detected, configure the host firewall to block connections by source-ip.
+
+### Questions
+1. If you had to deploy this program to hundreds of servers, what would be your preferred method? Why?
+4. What is the hardest technical problem or outage you've had to solve in your career? Explain what made it so difficult?
+
+# Scoring
```
Comment (klizhentas) [inline challenges/platform/devops.md:96]: @knisbet forgot one thing - we need a scorecard as well. We will add the scorecard in Lever and here. It will guide the interview team and the candidate and tell them how exactly are we going to evaluate their submission.

**Review by klizhentas — CHANGES_REQUESTED** [review]

File: challenges/platform/devops.md (hunk 10 above)
Comment (knisbet) [reply challenges/platform/devops.md:96]: I think this actually got removed from systems engineer since no one was really using it.

Comment (klizhentas) [reply challenges/platform/devops.md:96]: Talk about perception vs reality :) I totally missed that. I was under impression that it was very helpful for candidates to see how do we rank candidates submission and what areas we are evaluating. If not the scorecard, how do we tell them what aspects of the code/communication we are looking and not looking at?

Comment (knisbet) [reply challenges/platform/devops.md:96]: Tried to expand a bit on scoring in 663f3ed21ac7bf10a1a6c6e6e3fa31dca6404f1a. Instead of using a matrix like we used to use, a bit more high level with key focus areas.

**Review by klizhentas — APPROVED** [review]
